## Q1

In [ ]:
# (for windows os) download correct libsvm ver from https://www.lfd.uci.edu/~gohlke/pythonlibs/#libsvm
! pip install libsvm‑3.23‑cp37‑cp37m‑win_amd64.whl

In [1]:
import svm
import pandas as pd
import numpy as np
from svmutil import * # window OS ，pip install .whl file for libsvm alr

In [2]:
testf=r".\1\promoters\test.txt"
trainf=r".\1\promoters\training.txt"
testf=pd.read_csv(testf,header=None,sep=" ")
trainf=pd.read_csv(trainf,header=None,sep=" ")
test_label=list(testf[0])
train_label=list(trainf[0])
train=list(list(int(i[i.find(":")+1:]) for i in trainf.iloc[:,1:].values[j]) for j in range(len(trainf)))
test=list(list(int(i[i.find(":")+1:]) for i in testf.iloc[:,1:].values[j]) for j in range(len(testf)))

In [29]:
def svm(c,k):
    svm_model.predict = lambda self, x: svm_predict([0], [x], self)[0][0]
    prob = svm_problem(train_label,train)
    param = svm_parameter()
    param.kernel_type = LINEAR
    #param.C = c
    param.kernel_type=k
    m=svm_train(prob, param)
    #accuracy
    co=0
    for i in range(len(test)):
        if m.predict(test[i])==test_label[i]:
            co+=1
    accu=co/len(test)
    return accu

In [30]:
list(svm(5,i) for i in range(0,4))# scroll to last line of the output

Accuracy = 100% (1/1) (classification)
Accuracy = 0% (0/1) (classification)
Accuracy = 100% (1/1) (classification)
Accuracy = 100% (1/1) (classification)
Accuracy = 0% (0/1) (classification)
Accuracy = 0% (0/1) (classification)
Accuracy = 100% (1/1) (classification)
Accuracy = 0% (0/1) (classification)
Accuracy = 100% (1/1) (classification)
Accuracy = 0% (0/1) (classification)
Accuracy = 100% (1/1) (classification)
Accuracy = 0% (0/1) (classification)
Accuracy = 100% (1/1) (classification)
Accuracy = 100% (1/1) (classification)
Accuracy = 0% (0/1) (classification)
Accuracy = 100% (1/1) (classification)
Accuracy = 0% (0/1) (classification)
Accuracy = 100% (1/1) (classification)
Accuracy = 100% (1/1) (classification)
Accuracy = 0% (0/1) (classification)
Accuracy = 100% (1/1) (classification)
Accuracy = 100% (1/1) (classification)
Accuracy = 0% (0/1) (classification)
Accuracy = 0% (0/1) (classification)
Accuracy = 100% (1/1) (classification)
Accuracy = 100% (1/1) (classification)
Accuracy

[0.84375, 0.8125, 0.90625, 0.4375]

##### kernel = 2, accuracy =0.90625 is the best
## Q2 Q3  see attached pdf Q2

## Q4 .1

In [98]:
import timeit
code_to_test = """
import numpy as np
temp = np.random.rand(1000).astype(np.float64)
t=0
for i in temp:
    t*=i 
"""
elapsed_time = timeit.timeit(code_to_test, number=100)/100
print(elapsed_time)

0.00020914300000185904


In [101]:
code_to_test = """
import numpy as np
temp = np.log(np.random.rand(1000)).astype(np.float64)
t=0
for i in temp:
    t+=i 
"""
elapsed_time = timeit.timeit(code_to_test, number=100)/100
print(elapsed_time)

0.00018061900000247987


#### addition compared to multiplication is computationally cheaper.As above,test2 takes longer time than test2. stackoverflow problem occurs for multiplication, hence higher precision,hence memory and computation cost is required.

## 2.

In [103]:
import pandas as pd
testf=r".\4\diabetes_train.csv"
testf=pd.read_csv(testf,header=None)

In [104]:
testf.head(2) #1 label 20 attributes

0     1     2     3    4    5    6    7    8    9   ...   11   12   13  \
0 -1.0  0.54  0.68  0.65  1.0  0.0  0.0  0.0  1.0  0.0  ...  0.0  0.0  0.0   
1 -1.0  0.74  1.59  1.18  0.0  1.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   

    14   15   16   17   18   19   20  
0  0.0  1.0  0.0  0.0  0.0  1.0  0.0  
1  0.0  1.0  0.0  0.0  0.0  0.0  1.0  

[2 rows x 21 columns]

In [131]:
x, y = testf.drop(columns=0), testf.loc[:, 0][:,np.newaxis]
x['bias'] = np.ones((len(x), 1))

In [137]:
y.shape

(3000, 1)

In [138]:
def backward(x,y,theta):
    gradient=-(y*x)/1+np.exp(y*(x.dot(theta))) 
    gradient=gradient[:,np.newaxis]
    return gradient
def loss_func(x,y,theta,iteration):
    loss = np.sum(np.log(1+np.exp(np.multiply(-y,(x.values.dot(theta))))))
    acc = np.sum(np.sign(x.values.dot(theta)) == y)/ len(y)
    #print("iter "+str(iteration)+" : \n"+"loss: "+str(loss)+" accuracy: "+str(acc))
    print(f'iteration {iteration} \t loss {loss}, \t accuracy is {acc}')
    return loss,acc

In [139]:
import sys
from tqdm import tqdm_notebook, tqdm
losses = []
accs = []
lr = 0.1
theta = np.random.normal(0, 1, (x.shape[1], 1) )
iterations = 10000
idxs = np.random.choice(len(x), size=iterations)
weights = {}

for iteration, i in enumerate(tqdm_notebook(idxs, file=sys.stdout)):
    xi, yi = x.loc[i,:], y[i]
    gradient = backward(xi, yi, theta)
    theta -= lr * gradient
    if iteration % 100 == 0:
        loss, acc = loss_func(x, y, theta,iteration)
        weights[iteration] = theta
        losses.append(loss)
        accs.append(acc)

iteration 0 	 loss 1884.8405140375553, 	 accuracy is 0.6933333333333334


C:\Users\gaoyu\Anaconda3\lib\site-packages\ipykernel_launcher.py:7: RuntimeWarning: invalid value encountered in sign
  import sys


iteration 100 	 loss nan, 	 accuracy is 0.0
iteration 200 	 loss nan, 	 accuracy is 0.0
iteration 300 	 loss nan, 	 accuracy is 0.0
iteration 400 	 loss nan, 	 accuracy is 0.0
iteration 500 	 loss nan, 	 accuracy is 0.0
iteration 600 	 loss nan, 	 accuracy is 0.0
iteration 700 	 loss nan, 	 accuracy is 0.0
iteration 800 	 loss nan, 	 accuracy is 0.0
iteration 900 	 loss nan, 	 accuracy is 0.0
iteration 1000 	 loss nan, 	 accuracy is 0.0
iteration 1100 	 loss nan, 	 accuracy is 0.0
iteration 1200 	 loss nan, 	 accuracy is 0.0
iteration 1300 	 loss nan, 	 accuracy is 0.0
iteration 1400 	 loss nan, 	 accuracy is 0.0
iteration 1500 	 loss nan, 	 accuracy is 0.0
iteration 1600 	 loss nan, 	 accuracy is 0.0
iteration 1700 	 loss nan, 	 accuracy is 0.0
iteration 1800 	 loss nan, 	 accuracy is 0.0
iteration 1900 	 loss nan, 	 accuracy is 0.0
iteration 2000 	 loss nan, 	 accuracy is 0.0
iteration 2100 	 loss nan, 	 accuracy is 0.0
iteration 2200 	 loss nan, 	 accuracy is 0.0
iteration 2300 	 lo